targets/TARGET_ARM_SSG/TARGET_CM3DS_MPS2/i2c_api.c

```diff
@@ -0,0 +1,531 @@
+    }
+
+    i2c_delay(delay_value);
+    obj->i2c->CONTROLC = SDA;
+    i2c_delay(delay_value);
+    obj->i2c->CONTROLS = SCL;
+    i2c_delay(delay_value);
+    obj->i2c->CONTROLC = SCL;
+    i2c_delay(delay_value);
+
+}
+
+void i2c_init(i2c_t *obj, PinName sda, PinName scl)
+{
+    // determine the SPI to use
+    I2CName i2c_sda = (I2CName)pinmap_peripheral(sda, PinMap_I2C_SDA);
+    I2CName i2c_scl = (I2CName)pinmap_peripheral(scl, PinMap_I2C_SCL);
+    obj->i2c = (MPS2_I2C_TypeDef *)pinmap_merge(i2c_sda, i2c_scl);
+
+    if ((int)obj->i2c == NC) {
+        error("I2C pin mapping failed");
+    }
+
+    pinmap_pinout(sda, PinMap_I2C_SDA);
+    pinmap_pinout(scl, PinMap_I2C_SCL);
+
+    switch ((int)obj->i2c) {
+        case I2C_2:
+            GPIO_MAP[CMSDK_GPIO_SH0_SCL_I2C_GPIO_NUM]->ALTFUNCSET |=
+                                        (1<<CMSDK_GPIO_ALTFUNC_SH0_SCL_I2C_SET);
+            GPIO_MAP[CMSDK_GPIO_SH0_SDA_I2C_GPIO_NUM]->ALTFUNCSET |=
+                                        (1<<CMSDK_GPIO_ALTFUNC_SH0_SDA_I2C_SET);
+            break;
+        case I2C_3:
+            GPIO_MAP[CMSDK_GPIO_SH1_SCL_I2C_GPIO_NUM]->ALTFUNCSET |=
+                                        (1<<CMSDK_GPIO_ALTFUNC_SH1_SCL_I2C_SET);
+            GPIO_MAP[CMSDK_GPIO_SH1_SDA_I2C_GPIO_NUM]->ALTFUNCSET |=
+                                        (1<<CMSDK_GPIO_ALTFUNC_SH1_SDA_I2C_SET);
+            break;
+    }
+}
+
+int i2c_start(i2c_t *obj)
+{
+    int delay_value;
+    switch ((int)obj->i2c) {
+        case I2C_0: delay_value = TSC_TSU; break;
+        case I2C_1: delay_value = AAIC_TSU; break;
+        case I2C_2: delay_value = SHIELD_TSU; break;
+        case I2C_3: delay_value = SHIELD_TSU; break;
+    }
+
+    i2c_delay(delay_value);
+    obj->i2c->CONTROLS = SDA | SCL;
+    i2c_delay(delay_value);
+    obj->i2c->CONTROLC = SDA;
+    i2c_delay(delay_value);
+
+    return 0;
+}
+
+int i2c_start_tsc(i2c_t *obj)
+{
+    int delay_value;
+    switch ((int)obj->i2c) {
+        case I2C_0: delay_value = TSC_TSU; break;
+        case I2C_1: delay_value = AAIC_TSU; break;
+        case I2C_2: delay_value = SHIELD_TSU; break;
+        case I2C_3: delay_value = SHIELD_TSU; break;
+    }
+
+    i2c_delay(delay_value);
+    obj->i2c->CONTROLC = SDA;
+    i2c_delay(delay_value);
+    obj->i2c->CONTROLC = SCL;
+    i2c_delay(delay_value);
+
+    return 0;
+}
+
+int i2c_stop(i2c_t *obj)
+{
+    int delay_value;
+    switch ((int)obj->i2c) {
+        case I2C_0: delay_value = TSC_TSU; break;
+        case I2C_1: delay_value = AAIC_TSU; break;
+        case I2C_2: delay_value = SHIELD_TSU; break;
+        case I2C_3: delay_value = SHIELD_TSU; break;
+    }
+    // Actual stop bit
+    i2c_delay(delay_value);
+    obj->i2c->CONTROLC = SDA;
+    i2c_delay(delay_value);
+    obj->i2c->CONTROLS = SCL;
+    i2c_delay(delay_value);
+    obj->i2c->CONTROLS = SDA;
+    i2c_delay(delay_value);
+
+    return 0;
+}
+
+
+
+void i2c_frequency(i2c_t *obj, int hz) {
```

Comment:
Why is this function empty? Does the target support setting the I2C frequency?

@0xc0170 If the target doesn't support setting an I2C frequency, should it throw a runtime error/assert?